r1cs/gadgets/crypto/src/prf/ripemd160.rs

```diff
@@ -0,0 +1,458 @@
+];
+
+/// Apply the proper round function, selected with round_idx, to inputs a, b and c
+fn apply_round_function<ConstraintF, CS>(
+    cs:        CS,
+    a:         &UInt32,
+    b:         &UInt32,
+    c:         &UInt32,
+    round_idx: usize,
+) -> Result<UInt32, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>
+{
+    // Select and return the proper round function according to round idx
+    let (tri_fn, circ_fn): (fn(u32, u32, u32) -> u32, fn(&mut CS, usize, &Boolean, &Boolean, &Boolean) -> Result<Boolean, SynthesisError>) = match round_idx {
+        // f(j, a, b, c) = a XOR b XOR c (0 <= j <= 15)
+        round_idx if round_idx <= 15 => {
+            (
+                |a, b, c| a ^ b ^ c,
+                |cs, i, a, b, c| {
+                    let result = Boolean::xor(cs.ns(|| format!("A XOR B {}", i)), &a, &b)?;
+                    Boolean::xor(cs.ns(|| format!("A XOR B XOR C {}", i)), &result,&c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (a AND b) OR (NOT(a) AND c)  (16 <= j <= 31)
+        // Note: It's the same as the sha256_ch function e.g. (a AND b) XOR (NOT(a) AND c)
+        // since the two logical expressions have the same truth table
+        round_idx if round_idx >= 16 && round_idx <= 31 => {
+            (
+                |a, b, c| (a & b) | ((!a) & c),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("ch {}", i)), a, b, c)
+            )
+        },
+        // f(j, a, b, c) = (a OR NOT(b)) XOR c (32 <= j <= 47)
+        round_idx if round_idx >= 32 && round_idx <= 47 => {
+            (
+                |a, b, c| (a | (!b)) ^ c,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("A OR NOT B {}", i)), &a, &b.not())?;
+                    Boolean::xor(cs.ns(|| format!("(A OR NOT B) XOR C {}", i)), &t, &c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (c AND a) OR (NOT(c) AND b)  (48 <= j <= 63)
+        // Note: It's the same as the second round function, but with permuted variables
+        round_idx if round_idx >= 48 && round_idx <= 63 => {
+            (
+                |a, b, c| (c & a) | ((!c) & b),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("permuted ch {}", i)), c, a, b)
+            )
+        },
+        // f(j, a, b, c) = (b OR NOT(c)) XOR a (64 <= j <= 79)
+        // Note: It's the same as the third round function but with permuted variables
+        round_idx if round_idx >= 64 && round_idx <= 79 => {
+            (
+                |a, b, c| (b | (!c)) ^ a,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("B OR NOT C {}", i)), &b, &c.not())?;
+                    Boolean::xor(cs.ns(|| format!("(B OR NOT C) XOR a {}", i)), &t, &a)
+                }
+            )
+        },
+        _ => unreachable!()
+    };
+
+    triop(cs, a, b, c, tri_fn, circ_fn)
```

Comment:
What does it mean?

Reply:
I select the operands, primitive function and gadget function to apply to the operands depending on the round number. Then I call the function that actually executes it and put the constraints for it